Allow determining if an SVG value has been specified to enable animation

Ahmad-S792 opened the pull request:
#### a6a8ca738974cb312892566f4d9c1f824056bc03
<pre>
Allow determining if an SVG value has been specified to enable animation
<a href="https://bugs.webkit.org/show_bug.cgi?id=269481">https://bugs.webkit.org/show_bug.cgi?id=269481</a>
<a href="https://rdar.apple.com/problem/123457088">rdar://problem/123457088</a>

Reviewed by Nikolas Zimmermann.

When collecting attributes for pattern/gradient rendering, hasAttribute()
was used to check whether a property value should be included. This only
checks for the presence of a DOM attribute on the element, missing the case
where a value is produced solely by SMIL animation (animVal without a
corresponding baseVal attribute). The animated value was correctly computed
and reflected in the DOM API (animVal), but the rendering codepath never
read it.

Fix this by adding SVGElement::hasAttributeOrIsAnimatingProperty() which
combines hasAttribute() with a new registry query that checks whether the
animated property for a given attribute is currently being driven by an
animator. The plumbing goes through SVGMemberAccessor::isAnimating() →
SVGAnimatedPropertyAccessor (delegates to SVGAnimatedPropertyBase::isAnimating())
→ SVGPropertyOwnerRegistry::isAnimatingProperty() → SVGElement convenience
method. This preserves attribute inheritance semantics for paint servers
referencing other paint servers via href, since the check is false when
neither an attribute nor an animation is present.

Additionally, fix SVGFitToViewBox::hasValidViewBox() which suffered from a
similar issue: m_isViewBoxValid is only set when the viewBox DOM attribute
is parsed, so an animated viewBox without an explicit attribute was always
considered invalid. Now when the viewBox property is being animated, the
validity check bypasses the m_isViewBoxValid flag and directly validates
the animated value&apos;s dimensions.

Affects &lt;pattern&gt;, &lt;linearGradient&gt;, and &lt;radialGradient&gt; — all attributes
collected for rendering are now covered and fixes for both LegacySVG and
LBSE.

Inspired by (&amp; Tests): <a href="https://chromium.googlesource.com/chromium/blink/+/9102ba31c8dbaf5bd47d67a4c6450fbbaf75d43d">https://chromium.googlesource.com/chromium/blink/+/9102ba31c8dbaf5bd47d67a4c6450fbbaf75d43d</a>

* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::hasAttributeOrIsAnimatingProperty const):
* Source/WebCore/svg/SVGElement.h:
* Source/WebCore/svg/SVGFitToViewBox.h:
(WebCore::SVGFitToViewBox::hasValidViewBox):
* Source/WebCore/svg/SVGLinearGradientElement.cpp:
(WebCore::setGradientAttributes):
* Source/WebCore/svg/SVGPatternElement.cpp:
(WebCore::SVGPatternElement::collectPatternAttributes const):
* Source/WebCore/svg/SVGRadialGradientElement.cpp:
(WebCore::setGradientAttributes):
* Source/WebCore/svg/properties/SVGAnimatedPropertyAccessor.h:
* Source/WebCore/svg/properties/SVGAnimatedPropertyPairAccessor.h:
* Source/WebCore/svg/properties/SVGMemberAccessor.h:
(WebCore::SVGMemberAccessor::isAnimating const):
* Source/WebCore/svg/properties/SVGPropertyOwnerRegistry.h:
* Source/WebCore/svg/properties/SVGPropertyRegistry.h:

&gt; Add Test Case &amp; Expectations:
NOTE: Due to gradient noise, all gradient tests have `fuzziness` meta tag.
* LayoutTests/svg/animations/no-attr-radialgradient-spreadmethod.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-spreadmethod-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-r.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-r-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-gradientunits.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-gradientunits-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-gradienttransform.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-gradienttransform-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fy.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fy-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fx.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fx-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fr.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fr-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-cy.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-cy-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-cx.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-cx-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-y.svg:
* LayoutTests/svg/animations/no-attr-pattern-y-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-x.svg:
* LayoutTests/svg/animations/no-attr-pattern-x-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-width.svg:
* LayoutTests/svg/animations/no-attr-pattern-width-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-viewbox.svg:
* LayoutTests/svg/animations/no-attr-pattern-viewbox-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-patternunits.svg:
* LayoutTests/svg/animations/no-attr-pattern-patternunits-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-patterntransform.svg:
* LayoutTests/svg/animations/no-attr-pattern-patterntransform-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-patterncontentunits.svg:
* LayoutTests/svg/animations/no-attr-pattern-patterncontentunits-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-par.svg:
* LayoutTests/svg/animations/no-attr-pattern-par-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-height.svg:
* LayoutTests/svg/animations/no-attr-pattern-height-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-y2.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-y2-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-y1.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-y1-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-x2.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-x2-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-x1.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-x1-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-spreadmethod.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-spreadmethod-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-gradientunits.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-gradientunits-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-gradienttransform.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-gradienttransform-expected.svg:

Canonical link: <a href="https://commits.webkit.org/310257@main">https://commits.webkit.org/310257@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65a1b84a1f0f220c232213aba97f77ca0eb2a9e6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/153077 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/25859 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/19457 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/161821 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/106535 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/ed4b70c6-3130-4d2f-8c87-9d7f36fc7393) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/154950 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/26386 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/26164 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/118321 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/83773 "3 flakes 6 failures") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/11d99040-7845-4e95-a3c7-22bb416c6dc8) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/156036 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/20556 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/137422 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/99034 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/9e3e4f99-fedf-4da3-8a91-4b42e965e839/91b5c611-c1e4-4837-93e5-b40e2717f1d7) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/19630 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/17572 "Found 1 new API test failure: TestWebKitAPI.WKWebExtensionAPINamespace.NoWebNavigationObjectWithoutPermission (failure)") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/9657 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/129283 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/167/builds/15295 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/164295 "Built successfully") | | 
| | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/7431 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/16889 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/126383 "Passed tests") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/25656 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/21610 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/126541 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/34367 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/25658 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/137091 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/82302 "Built successfully") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/21492 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/13870 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/25274 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/89561 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/24967 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/25125 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/25026 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/svg/SVGPatternElement.cpp:
```diff
@@ -157,31 +157,31 @@ RenderPtr<RenderElement> SVGPatternElement::createElementRenderer(RenderStyle&&
 void SVGPatternElement::collectPatternAttributes(PatternAttributes& attributes) const
 {
-    if (!attributes.hasX() && hasAttribute(SVGNames::xAttr))
+    if (!attributes.hasX() && (isAnimatedAttribute(SVGNames::xAttr) || hasAttribute(SVGNames::xAttr)))
```
Hmm, wouldn't a helper method does check both be more convenient?

Ahmad-S792 replied:
Yes - I will get back to this - I think I need to introduce helper function because `isAnimateAttributes` lead to failures and with using both `||` still leads to failures. Let me rebase this first and push and then do local testing and fix this.

nikolaszimmermann reviewed (commented):
Looks fine. I wonder if we have more cases of wrong hasAttribute usage? Shall we make those tests WPT tests?

nikolaszimmermann commented on LayoutTests/svg/animations/no-attr-lineargradient-gradienttransform.svg:
```diff
@@ -0,0 +1,10 @@
+  <linearGradient id="gradient">
+    <stop stop-color="#008000" offset="0.5"/>
+    <stop stop-color="#ff0000" offset="0.5"/>
+    <animateTransform attributeName="gradientTransform" type="scale" dur="0.001s" from="2" to="2" fill="freeze"/>
```
I hope this is not introducing flakiness -- the SVG anim tests are notoriously flaky already, because we're not properly controlling the SMIL timeline... A small duration like this looks suspicious - I wonder if there's a better way to test this. How do other WPT+SVG SMIL tests handle these situations?

Ahmad-S792 replied:
WPT+SVG mostly are testharness + retests. I tink `stress test` bot run these through 100 iterations and didn't flag any for flakiness, so I think it would be OK. @nikolaszimmermann

nikolaszimmermann replied:
Ok, let's try it.

nikolaszimmermann commented on Source/WebCore/svg/SVGFitToViewBox.h:
```diff
@@ -56,7 +56,13 @@ class SVGFitToViewBox {
+    {
+        auto& viewBox = this->viewBox();
+        if (m_viewBox->isAnimating())
+            return viewBox.width() >= 0 && viewBox.height() >= 0;
```
Do we need an equivalent of the `m_isViewBoxValid` check while animating, or is the condition already  viewBox.width>0 || viewBox.height>0 sufficient?

Ahmad-S792 replied:
Let me look but from quick, no harm in adding additional check.

nikolaszimmermann replied:
Well we cannot use m_isViewBoxValid directly, that would refer to the non-animated variant of the viewBox. I think it's only there to differentiate between the "valid" viewBox `viewBox="0 0 0 0"` vs. viewBox not set at all.

Therefore I think it's implicitly encoded  within `m_viewBox->isAnimating()` for the SMIL-animVal case -- it's definately set if we take that branch, so your patch is fine as is.

Ahmad-S792 replied:
I was just about to switch to explore it. Thanks!

nikolaszimmermann reviewed (commented):
LGTM, just a single question remaining.

Ahmad-S792 committed:
Allow determining if an SVG value has been specified to enable animation
https://bugs.webkit.org/show_bug.cgi?id=269481
rdar://problem/123457088

Reviewed by Nikolas Zimmermann.

When collecting attributes for pattern/gradient rendering, hasAttribute()
was used to check whether a property value should be included. This only
checks for the presence of a DOM attribute on the element, missing the case
where a value is produced solely by SMIL animation (animVal without a
corresponding baseVal attribute). The animated value was correctly computed
and reflected in the DOM API (animVal), but the rendering codepath never
read it.

Fix this by adding SVGElement::hasAttributeOrIsAnimatingProperty() which
combines hasAttribute() with a new registry query that checks whether the
animated property for a given attribute is currently being driven by an
animator. The plumbing goes through SVGMemberAccessor::isAnimating() →
SVGAnimatedPropertyAccessor (delegates to SVGAnimatedPropertyBase::isAnimating())
→ SVGPropertyOwnerRegistry::isAnimatingProperty() → SVGElement convenience
method. This preserves attribute inheritance semantics for paint servers
referencing other paint servers via href, since the check is false when
neither an attribute nor an animation is present.

Additionally, fix SVGFitToViewBox::hasValidViewBox() which suffered from a
similar issue: m_isViewBoxValid is only set when the viewBox DOM attribute
is parsed, so an animated viewBox without an explicit attribute was always
considered invalid. Now when the viewBox property is being animated, the
validity check bypasses the m_isViewBoxValid flag and directly validates
the animated value's dimensions.

Affects <pattern>, <linearGradient>, and <radialGradient> — all attributes
collected for rendering are now covered and fixes for both LegacySVG and
LBSE.

Inspired by (& Tests): https://chromium.googlesource.com/chromium/blink/+/9102ba31c8dbaf5bd47d67a4c6450fbbaf75d43d

* Source/WebCore/svg/SVGElement.cpp:
(WebCore::SVGElement::hasAttributeOrIsAnimatingProperty const):
* Source/WebCore/svg/SVGElement.h:
* Source/WebCore/svg/SVGFitToViewBox.h:
(WebCore::SVGFitToViewBox::hasValidViewBox):
* Source/WebCore/svg/SVGLinearGradientElement.cpp:
(WebCore::setGradientAttributes):
* Source/WebCore/svg/SVGPatternElement.cpp:
(WebCore::SVGPatternElement::collectPatternAttributes const):
* Source/WebCore/svg/SVGRadialGradientElement.cpp:
(WebCore::setGradientAttributes):
* Source/WebCore/svg/properties/SVGAnimatedPropertyAccessor.h:
* Source/WebCore/svg/properties/SVGAnimatedPropertyPairAccessor.h:
* Source/WebCore/svg/properties/SVGMemberAccessor.h:
(WebCore::SVGMemberAccessor::isAnimating const):
* Source/WebCore/svg/properties/SVGPropertyOwnerRegistry.h:
* Source/WebCore/svg/properties/SVGPropertyRegistry.h:

> Add Test Case & Expectations:
NOTE: Due to gradient noise, all gradient tests have `fuzziness` meta tag.
* LayoutTests/svg/animations/no-attr-radialgradient-spreadmethod.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-spreadmethod-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-r.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-r-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-gradientunits.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-gradientunits-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-gradienttransform.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-gradienttransform-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fy.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fy-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fx.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fx-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fr.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-fr-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-cy.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-cy-expected.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-cx.svg:
* LayoutTests/svg/animations/no-attr-radialgradient-cx-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-y.svg:
* LayoutTests/svg/animations/no-attr-pattern-y-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-x.svg:
* LayoutTests/svg/animations/no-attr-pattern-x-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-width.svg:
* LayoutTests/svg/animations/no-attr-pattern-width-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-viewbox.svg:
* LayoutTests/svg/animations/no-attr-pattern-viewbox-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-patternunits.svg:
* LayoutTests/svg/animations/no-attr-pattern-patternunits-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-patterntransform.svg:
* LayoutTests/svg/animations/no-attr-pattern-patterntransform-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-patterncontentunits.svg:
* LayoutTests/svg/animations/no-attr-pattern-patterncontentunits-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-par.svg:
* LayoutTests/svg/animations/no-attr-pattern-par-expected.svg:
* LayoutTests/svg/animations/no-attr-pattern-height.svg:
* LayoutTests/svg/animations/no-attr-pattern-height-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-y2.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-y2-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-y1.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-y1-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-x2.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-x2-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-x1.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-x1-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-spreadmethod.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-spreadmethod-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-gradientunits.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-gradientunits-expected.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-gradienttransform.svg:
* LayoutTests/svg/animations/no-attr-lineargradient-gradienttransform-expected.svg:

Canonical link: https://commits.webkit.org/310257@main
